ReactiveCocoaTests/Swift/SignalSpec.swift

```diff
@@ -14,30 +14,154 @@ import ReactiveCocoa
+				}
+				
+				var errored = false
+				
+				signal?.observe(error: { _ in errored = true })
+				
+				expect(errored).to(beFalsy())
+				expect(signal).toNot(beNil())
+				
+				testScheduler.run()
+				
+				expect(errored).to(beTruthy())
+				expect(signal).to(beNil())
+			}
+
+			it("should deallocate after completing") {
+				weak var signal: Signal<AnyObject, NoError>? = Signal { observer in
+					testScheduler.schedule {
+						sendCompleted(observer)
+					}
+					return nil
+				}
+				
+				var completed = false
+				
+				signal?.observe(completed: { completed = true })
+				
+				expect(completed).to(beFalsy())
+				expect(signal).toNot(beNil())
+				
+				testScheduler.run()
+				
+				expect(completed).to(beTruthy())
+				expect(signal).to(beNil())
+			}
+
+			it("should forward events to observers") {
+				let numbers = [ 1, 2, 5 ]
+				
+				let signal: Signal<Int, NoError> = Signal { observer in
+					testScheduler.schedule {
+						for number in numbers {
+							sendNext(observer, number)
+						}
+						sendCompleted(observer)
+					}
+					return nil
+				}
+				
+				var fromSignal: [Int] = []
+				var completed = false
+				
+				signal.observe(next: { number in
+					fromSignal.append(number)
+				}, completed: {
+					completed = true
+				})
+				
+				expect(completed).to(beFalsy())
+				expect(fromSignal).to(beEmpty())
+				
+				testScheduler.run()
+				
+				expect(completed).to(beTruthy())
+				expect(fromSignal).to(equal(numbers))
+			}
+
+			it("should dispose of returned disposable upon error") {
+				let disposable = SimpleDisposable()
+				
+				let signal: Signal<AnyObject, TestError> = Signal { observer in
+					testScheduler.schedule {
+						sendError(observer, TestError.Default)
+					}
+					return disposable
+				}
+				
+				var errored = false
+				
+				signal.observe(error: { _ in errored = true })
+				
+				expect(errored).to(beFalsy())
+				expect(disposable.disposed).to(beFalsy())
+				
+				testScheduler.run()
+				
+				expect(errored).to(beTruthy())
+				expect(disposable.disposed).to(beTruthy())
+			}
+
+			it("should dispose of returned disposable upon completion") {
+				let disposable = SimpleDisposable()
+				
+				let signal: Signal<AnyObject, NoError> = Signal { observer in
+					testScheduler.schedule {
+						sendCompleted(observer)
+					}
+					return disposable
+				}
+				
+				var completed = false
+				
+				signal.observe(completed: { completed = true })
+				
+				expect(completed).to(beFalsy())
+				expect(disposable.disposed).to(beFalsy())
+				
+				testScheduler.run()
+				
+				expect(completed).to(beTruthy())
+				expect(disposable.disposed).to(beTruthy())
 			}
 		}
 
 		describe("Signal.pipe") {
-			pending("should keep signal alive if not terminated") {
+			
+			it("should keep signal alive if not terminated") {
+				weak var signal = Signal<(), NoError>.pipe().0
+				
+				expect(signal).toNot(beNil())
 			}
 
 			pending("should deallocate after erroring") {
```

Comment:
You may need to put some logic inside a closure, to put explicit bounds on the lifetime of some objects. See [the old tests](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/swift-before-signal-producers/ReactiveCocoaTests/Swift/HotSignalSpec.swift#L18-L39) for some examples.